### PR TITLE
refactor: remove cita-types, hashable, rlp dependencies from merklehash

### PR DESCRIPTION
**Pull request by u2**
1. remove `cita-types`, `hashable`, `rlp` dependencies from merklehash
2. add another crate cita-merklehash associated with the `cita-types`, `cita-hash`, `cita-serde` and `cita-merge`.

TODO: 
1. Add a TreeBuilder in cita-merklehash
2. Rename merklehash to static-merkle
3. fix the test, merkelhash should build without any features.

https://github.com/cryptape/cita-common/issues/170

### CODE REVIEW

File: merklehash/src/lib.rs [hunk 1]
```diff
@@ -18,33 +18,30 @@
...
-extern crate rlp;
-#[macro_use]
-extern crate rlp_derive;
-
-use cita_types::H256;
-use hashable::{Hashable, HASH_NULL_RLP};
-use rlp::RlpStream;
+pub trait Merge<T> {
+    fn merge(&self, left: &T, right: &T) -> T;
+    fn null(&self) -> T;
```
Comment (u2) [inline merklehash/src/lib.rs]: How to remove self?

File: merklehash/src/lib.rs [hunk 2]
```diff
@@ -89,25 +86,25 @@ impl Tree {
 
                 // Insert the input into the merkle tree.
                 for i in 0..nodes_size_of_lowest_row {
-                    nodes[first_index_in_lowest_row + i] = input[i]
+                    nodes[first_index_in_lowest_row + i] = input[i].clone()
```
Comment (u2) [inline merklehash/src/lib.rs:90]: Before there is no clone, does the H256 have copy trait?

File: merklehash/src/lib.rs [hunk 3]
```diff
@@ -117,50 +114,44 @@ impl Tree {
...
             input_size: input_size,
+            merge: merge,
         }
     }
 
-    pub fn from_bytes<I>(input: I) -> Self
-    where
-        I: IntoIterator<Item = Vec<u8>>,
-    {
-        Self::from_hashes(input.into_iter().map(|v| v.crypt_hash()).collect())
```
Comment (u2) [inline merklehash/src/lib.rs:127]: Remove the from_bytes from the Tree, or we add a hash trait for it.

File: merklehash/src/lib.rs [hunk 4]
```diff
@@ -117,50 +114,44 @@ impl Tree {
...
+                        hash: self.nodes[i].clone(),
                     })
                     .collect(),
             )
         })
     }
 }
 
-impl Proof {
-    pub fn verify(&self, root_hash: H256, data_hash: H256) -> bool {
```
Comment (u2) [inline merklehash/src/lib.rs:150]: rename to root, data

File: merklehash/src/lib.rs (hunk 1 above)
Comment (keroro520) [reply merklehash/src/lib.rs]: Trait allow defining function inside it. So remove `&self` directly.

File: merklehash/src/lib.rs (hunk 2 above)
Comment (chaoticlonghair) [reply merklehash/src/lib.rs:90]: Yes, and clippy will not allow `clone()` here.

**Comment by u2** [comment]
Forgot to commit some files, updated.

**Comment by u2** [comment]
@yangby-cryptape @keroro520 , updated, please review.

**Review by chaoticlonghair — DISMISSED** [review]

File: libproto/src/lib.rs [hunk 5]
```diff
@@ -389,7 +392,9 @@ impl CompactBlockBody {
     }
 
     pub fn transactions_root(&self) -> H256 {
-        merklehash::Tree::from_hashes(self.transaction_hashes().clone()).get_root_hash()
+        Tree::from_hashes(self.transaction_hashes().clone(), merge, HASH_NULL)
+            .get_root_hash()
+            .clone()
```
Comment (chaoticlonghair) [inline libproto/src/lib.rs]: This `clone()` can be removed.
And other `clone()` can be removed, too.

File: merklehash/src/lib.rs (hunk 2 above)
Comment (u2) [reply merklehash/src/lib.rs:90]: ok

File: libproto/src/lib.rs (hunk 5 above)
Comment (u2) [reply libproto/src/lib.rs]: updated

**Review by chaoticlonghair — APPROVED** [review]

**Pull request merged**